File: simulator/spells/magic_missile.py
from functools import cache

from cachetools import cached
from cachetools.keys import hashkey

from ..actions.action_types import BonusAction
from ..battle_map import Map, map_position_toggled_cache, map_toggled_cache_with_key, \
    _get_free_coords_in_cartesian_range
from ..spells.spell import SpellStats
from ..misc import DamageType, Visibility
from ..conditions import Conditions, is_affected_by_any, is_affected_by, get_swallower
from ..actions.actoid import Actoid, ActoidFlags
from ..threat_utils import mean_dmg_auto_hit
from ..threat_interfaces import DirectThreat
from ..factory_interfaces import DirectThreatFactory
from itertools import combinations_with_replacement
import logging
from ..utils.roll_types import RollType

logger = logging.getLogger("Encounterra")

class MagicMissileFactory(DirectThreatFactory):
    level = 1
    range = SpellStats.Range.FEET_120.value
    target = SpellStats.Target.THREE_CREATURES
    duration = SpellStats.Duration.INSTANTANEOUS
    concentration = False
    type = SpellStats.Type.HARMFUL
    dc = None
    dmg_type = DamageType.Force

    def __init__(self, action_type, caster, resource):
        super().__init__()
        self.action_type = action_type  # MAGIC_MISSILE, QUICKENED_MAGIC_MISSILE
        self.dmg_dice = ((1, 4),)
        self.dmg_bonus = 1
        self.combatant = caster
        self.resource = resource

    def __str__(self):
        """
        Important for FSM building
        """
        return "MagicMissileFactory"

    def get_ability_name(self):
        return "Magic Missile"


    def get_quickened_kwargs(self):
        return {'caster': self.combatant, 'resource': self.resource}

    def get_eligible_targets(self):
        swallower = get_swallower(self.combatant)
        if swallower:
            return [swallower, swallower, swallower]
        # Range is so big that it doesn't matter
        return combinations_with_replacement([e for e in Map.get().get_non_swallowed_enemies(self.combatant)], 3)

    def create_all(self, previous_action_in_dag=None):
        targets = self.get_eligible_targets()
        return [MagicMissile(t, self) for t in targets]

    def create(self, targets):
        return MagicMissile(targets, self)

    def calculate_threat_to_target(self, target, **kwargs):
        battle_map = Map.get()
        if battle_map.get_cartesian_distance_combatants(self.combatant, target) <= MagicMissileFactory.range:
            ret = 3 * (mean_dmg_auto_hit(self.dmg_dice, target.is_resistant_to(MagicMissileFactory.dmg_type)) + self.dmg_bonus)
            return ret
        else:
            return 0

    def calculate_threat_to_target_delta(self, target, modifiers, *args, **kwargs):
        """
        Calculates the threat delta of the factory to a specific target given stat modifications.
        This is useful calculating the potential reduction of threat_in caused by abilities of enemies, e.g. advantage on saving throw
        against fireball or bane on attack rolls etc.
        """
        return 0

    def calculate_max_threat(self):
        if get_swallower(self.combatant):
            return 0  # Must be able to see
        targets = [e for e in Map.get().get_non_swallowed_enemies(self.combatant)]
        for t in targets:
            threat = self.calculate_threat_to_target(t)
            # We just need one enemy within range which assures we can deal the damage (which is target-agnostic)
            if threat:
                return threat
        return 0


class MagicMissile(Actoid, DirectThreat):

    def __init__(self, targets, factory, **kwargs):
        Actoid.__init__(self, ActoidFlags.IS_SPELL)
        self.targets = targets
        self.factory = factory
        self.empowered = kwargs.get("empowered", False)
        self.roll_type = RollType.STRAIGHT

    def __str__(self):
        if self.targets[0] is self.targets[1] and self.targets[1] is self.targets[2]:
            return (
                "Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + f"Magic Missile on 3x{self.targets[0]}"
        elif self.targets[0] is self.targets[1]:
            return (
                "Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + f"Magic Missile on 2x{self.targets[0]} and {self.targets[2]}"
        elif self.targets[1] is self.targets[2]:
            return (
                "Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + f"Magic Missile on 2x{self.targets[1]} and {self.targets[0]}"
        elif self.targets[0] is self.targets[2]:
            return (
                "Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + f"Magic Missile on 2x{self.targets[0]} and {self.targets[1]}"
        return (
            "Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + f"Magic Missile on {self.targets[0]}, {self.targets[1]} and {self.targets[2]}"

    def shorthand_str(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_MAGIC_MISSILE else "") + "Magic Missile"

    @map_position_toggled_cache
    def calculate_threat(self, **kwargs):
        dmg_acc = mean_dmg_auto_hit(self.factory.dmg_dice, self.targets[0].is_resistant_to(MagicMissileFactory.dmg_type)) + self.factory.dmg_bonus
        dmg_acc += mean_dmg_auto_hit(self.factory.dmg_dice, self.targets[1].is_resistant_to(MagicMissileFactory.dmg_type)) + self.factory.dmg_bonus
        dmg_acc += mean_dmg_auto_hit(self.factory.dmg_dice, self.targets[2].is_resistant_to(MagicMissileFactory.dmg_type)) + self.factory.dmg_bonus
        return dmg_acc

    def clear_cache(self):
        self.calculate_threat.cache_clear()
        #self.get_eligible_coords.cache_clear()

    def calculate_threat_delta(self, modifiers, *args, **kwargs):
        return 0

    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        if get_swallower(self.factory.combatant):
            return None  # Must be able to see
        curr_coord = tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])
        if not is_affected_by_any(self.factory.combatant, Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
            coords_for_first = set(_get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.targets[0]).get(),
                distances,
                inflate_to_dist=self.factory.combatant.size.value,
                rng=MagicMissileFactory.range,
                combatant_id=self.factory.combatant.id))
            coords_for_second = set(_get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.targets[1]).get(),
                distances,
                inflate_to_dist=self.factory.combatant.size.value,
                rng=MagicMissileFactory.range,
                combatant_id=self.factory.combatant.id))
            coords_for_third = set(_get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.targets[2]).get(),
                distances,
                inflate_to_dist=self.factory.combatant.size.value,
                rng=MagicMissileFactory.range,
                combatant_id=self.factory.combatant.id))
            free_coords_in_range = coords_for_third.intersection(coords_for_first.intersection(coords_for_second))

            return [coord for coord in free_coords_in_range if battle_map.visibility_dict_for_all_coords[coord][self.targets[0]] is not Visibility.NONE
                    and battle_map.visibility_dict_for_all_coords[coord][self.targets[1]] is not Visibility.NONE
                    and battle_map.visibility_dict_for_all_coords[coord][self.targets[2]] is not Visibility.NONE]
        elif battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.targets[0]) <= MagicMissileFactory.range \
            and battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.targets[1]) <= MagicMissileFactory.range \
            and battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.targets[2]) <= MagicMissileFactory.range \
            and battle_map.visibility_dict_for_all_coords[curr_coord][self.targets[0]] is not Visibility.NONE \
            and battle_map.visibility_dict_for_all_coords[curr_coord][self.targets[1]] is not Visibility.NONE \
                and battle_map.visibility_dict_for_all_coords[curr_coord][self.targets[2]] is not Visibility.NONE:
            return [curr_coord]
        return None

File: c++/spells/magic_missile.cpp
#include "spells/magic_missile.hpp"
#include "core/combatant.hpp"
#include "core/battle_map.hpp"
#include <algorithm>
#include <limits>

namespace enc
{
  MagicMissileFactory::MagicMissileFactory(AbilityType abilityType, Combatant *caster, Resource *resource)
      : DirectThreatFactory("MagicMissileFactory", "Magic Missile", caster, abilityType), _resource(resource)
  {}

  std::vector<std::array<Combatant *, 3>> MagicMissileFactory::getEligibleTargetSets() const
  {
    if(auto *swallower = _combatant->getSwallower())
      {
        return {{{swallower, swallower, swallower}}};
      }

    BattleMap &battleMap = BattleMap::getInstance();
    std::vector<Combatant *> enemies = battleMap.getNonSwallowedEnemiesWithinRadius(_combatant, static_cast<int>(MagicMissileFactory::range));
    std::vector<std::array<Combatant *, 3>> targetSets;
    for(size_t i = 0; i < enemies.size(); ++i)
      {
        for(size_t j = i; j < enemies.size(); ++j)
          {
            for(size_t k = j; k < enemies.size(); ++k)
              {
                targetSets.push_back({enemies[i], enemies[j], enemies[k]});
              }
          }
      }
    return targetSets;
  }

  std::vector<std::shared_ptr<Actoid>> MagicMissileFactory::createAll(void *previousActionInDag)
  {
    auto targetSets = getEligibleTargetSets();
    std::vector<std::shared_ptr<Actoid>> result;
    result.reserve(targetSets.size());
    for(const auto &targets : targetSets)
      {
        result.push_back(std::make_shared<MagicMissile>(targets, *this));
      }
    return result;
  }

  std::shared_ptr<Actoid> MagicMissileFactory::create(void *target)
  {
    return std::make_shared<MagicMissile>(*static_cast<std::array<Combatant *, 3> *>(target), *this);
  }

  double MagicMissileFactory::calculateThreatToTarget(Combatant *target, const Kwargs &kwargs) const
  {
    BattleMap &battleMap = BattleMap::getInstance();
    if(target->getSwallower() || battleMap.getCartesianDistanceCombatants(*_combatant, *target) > static_cast<int>(MagicMissileFactory::range))
      {
        return 0.0;
      }
    if(target->isImmuneTo(MagicMissileFactory::dmgType))
      {
        return 0.0;
      }
    double dartThreat = avgRoll(MagicMissileFactory::dmgDice) + MagicMissileFactory::dmgBonus;
    if(target->isResistantTo(MagicMissileFactory::dmgType))
      {
        dartThreat /= 2.0;
      }
    return std::min(static_cast<double>(target->getCurrentHp()), 3.0 * dartThreat);
  }

  double MagicMissileFactory::calculateThreatToTargetDelta(Combatant *target, const ThreatModifiers &modifiers) const { return 0.0; }

  double MagicMissileFactory::calculateMaxThreat() const
  {
    double maxThreat = 0.0;
    for(const auto &targets : getEligibleTargetSets())
      {
        maxThreat = std::max(maxThreat, MagicMissile(targets, *this).calculateThreat({}));
      }
    return maxThreat;
  }

  std::string MagicMissile::toString() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_MAGIC_MISSILE) ? "Quickened " : "";
    return prefix + "Magic Missile on " + _targets[0]->_name + ", " + _targets[1]->_name + ", and " + _targets[2]->_name;
  }

  std::string MagicMissile::shorthandStr() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_MAGIC_MISSILE) ? "Quickened " : "";
    return prefix + "Magic Missile";
  }

  double MagicMissile::calculateThreat(const Kwargs &kwargs)
  {
    double threat = 0.0;
    for(auto *target : _targets)
      {
        if(target->isImmuneTo(MagicMissileFactory::dmgType))
          {
            continue;
          }
        double dartThreat = avgRoll(MagicMissileFactory::dmgDice) + MagicMissileFactory::dmgBonus;
        if(target->isResistantTo(MagicMissileFactory::dmgType))
          {
            dartThreat /= 2.0;
          }
        threat += dartThreat;
      }
    return threat;
  }

  double MagicMissile::calculateThreatDelta(const ThreatModifiers &modifiers) const { return 0.0; }

  std::optional<CoordVector>
  MagicMissile::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    BattleMap &battleMap = BattleMap::getInstance();
    Combatant *caster = _factory._combatant;
    Coord currCoord = battleMap.getCombatantCoordinates(*caster).getRoot();
    if(caster->getSwallower())
      {
        return CoordVector{currCoord};
      }

    auto inRangeAtCurrent = [&]() {
      return std::all_of(_targets.begin(), _targets.end(), [&](Combatant *target) {
        return battleMap.getCartesianDistanceCombatants(*caster, *target) <= static_cast<int>(MagicMissileFactory::range);
      });
    };

    if(caster->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        return inRangeAtCurrent() ? std::optional<CoordVector>(CoordVector{currCoord}) : std::nullopt;
      }

    CoordVector coords = battleMap.getFreeCoordsInCartesianRange(battleMap.getCombatantCoordinates(*_targets[0]).get(), distances, caster->getSize(),
                                                                 static_cast<int>(MagicMissileFactory::range), caster->_instanceId);
    for(size_t i = 1; i < _targets.size(); ++i)
      {
        CoordVector targetCoords = battleMap.getFreeCoordsInCartesianRange(battleMap.getCombatantCoordinates(*_targets[i]).get(), distances,
                                                                           caster->getSize(), static_cast<int>(MagicMissileFactory::range),
                                                                           caster->_instanceId);
        coords.erase(std::remove_if(coords.begin(), coords.end(), [&targetCoords](const Coord &coord) {
                       return std::find(targetCoords.begin(), targetCoords.end(), coord) == targetCoords.end();
                     }),
                     coords.end());
      }
    return coords;
  }
}
